### PR TITLE
UIU-516 Spike: Select a Password Strength Meter

### PR DESCRIPTION
**Pull request by oleksii-popov**
Added Password Strength component with readme.md
Added translations for EN
Added password strength plugin to package.json
Changelog updated correspondingly

### CODE REVIEW

File: lib/PasswordStrength/PasswordStrength.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    return (
+      <div>
+        <FormattedMessage id={'stripes-smart-components.password-strength.' + output}/>
+      </div>
+    );
+  }
+}
+
+export default withStripes(PasswordStrength);
```
Comment (cherewaty) [inline lib/PasswordStrength/PasswordStrength.js]: It doesn't look like you're accessing anything on the `stripes` god object, so I don't think you need `withStripes`. You can just `export default PasswordStrength;`.

**Comment by doytch** [comment]
Never really looked into this lib so I'm just firing in the dark, but is it necessary to re-create the strengthTester, add the common passwords and trigraph map in every render call? Could we just do this in the constructor once?

Also, theoretically, if we did do that, we could just run the `check()` calls in a `getDerivedStateFromProps` if the password has changed.

Finally, just to be pedantic, can we rename the prop to `password`, especially since you rename it as such in `render`? I'll never remember `dataSource` 😞

File: lib/PasswordStrength/PasswordStrength.js (hunk 1 above)
Comment (oleksii-popov) [reply lib/PasswordStrength/PasswordStrength.js]: @cherewaty  thanks for your comment. 
I'm new to stripes platform. And I did not find a way to pass stripes to this component other then <code>withStripes</code> (I found it in another module). If you will describe how to pass <code>stripes</code> to the component, I will really appreciate.

Comment (zburke) [reply lib/PasswordStrength/PasswordStrength.js]: I think @cherewaty was suggesting that since you don't use `stripes` anywhere in this class, you don't need to wrap it with `withStripes`.

Comment (oleksii-popov) [reply lib/PasswordStrength/PasswordStrength.js]: It is needed for translation strings.

**Comment by oleksii-popov** [comment]
Added aria-live="polite"
Moved tester init and trigraphs to a constructor
Password strength check logic moved out of 'render'.
Removed 'withStripes'


@JohnC-80  @doytch please review
+ @VictorSoroka1 @TarasSpashchenko @OleksiiKuzminov @Igor-Gorchakov @natali-zaitseva

File: lib/PasswordStrength/PasswordStrength.js (hunk 1 above)
Comment (oleksii-popov) [reply lib/PasswordStrength/PasswordStrength.js]: Updated, please take a look. No more withStripes

**Comment by oleksii-popov** [comment]
@MikeTaylor, @zburke can you please take a look.

File: lib/PasswordStrength/PasswordStrength.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    this.output = 'weak'; // set default password hint to weak
+  }
+
+  componentDidUpdate(prevProps) {
+    // in case password updated and provided
+    if (this.props.password !== prevProps.password && !!this.props.password) {
+      const password = this.props.password;
+      const result = this.strengthTester.check(password);
+
+      switch (true) {
```
Comment (MikeTaylor) [inline lib/PasswordStrength/PasswordStrength.js]: The `switch (true)` definitely falls into the "too clever" category.

> "Everyone knows that debugging is twice as hard at writing a program in the first place.  So if you're as clever as you can be be when you write it, how will you ever debug it?" -- Kernighan & Pike, _The Elements of Programming Style_.

Please rephrase this as a regular sequence of `if`s such as programmers have been happily using since 1958.

File: lib/PasswordStrength/PasswordStrength.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+          this.output = 'reasonable';
+          break;
+      }
+    }
+  }
+
+  render() {
+    const password = this.props.password;
+    if (!password) {
+      return false;
```
Comment (MikeTaylor) [inline lib/PasswordStrength/PasswordStrength.js]: Does React use the `false` return-value? If not, it's more honest just to `return`.

File: lib/PasswordStrength/readme.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+## Introduction
+
+It is POC for password strength component. It requies additional styling, 
+tests and configuration in order to be completed.
+
+## Usage
+
+Include plugin via 
+`<PasswordStrength dataSource={DATA_STRING}></PasswordStrength>`
```
Comment (MikeTaylor) [inline lib/PasswordStrength/readme.md]: Shouldn't this be `password={DATA_STRING}`?

File: lib/PasswordStrength/readme.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import { formValues } from 'redux-form';
+
+getCurrentValues(key) {
+        const {stripes: {store}} = this.props;
+        const state = store.getState();
+        const values = getFormValues('Form')(state) || {};
+        return values[key] || false;
+}
+
+<PasswordStrength type="password-strength" dataSource={this.getCurrentValues("InputName")}>
```
Comment (MikeTaylor) [inline lib/PasswordStrength/readme.md]: Likewise, `password`?

File: lib/PasswordStrength/PasswordStrength.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    super(props);
+    this.strengthTester = new taiPasswordStrength.PasswordStrength();
+    this.strengthTester.addCommonPasswords(taiPasswordStrength.commonPasswords);
+    this.strengthTester.addTrigraphMap(taiPasswordStrength.trigraphs);
+    this.output = 'weak'; // set default password hint to weak
+  }
+
+  componentDidUpdate(prevProps) {
+    // in case password updated and provided
+    if (this.props.password !== prevProps.password && !!this.props.password) {
```
[the remark this answers is not in the document]
Comment (oleksii-popov) [reply lib/PasswordStrength/PasswordStrength.js]: Thank you, updated

File: lib/PasswordStrength/PasswordStrength.js (hunk 2 above)
Comment (oleksii-popov) [reply lib/PasswordStrength/PasswordStrength.js]: As it is POC - I presented the main idea of working with strengthTester results.
In a final version, there should be more check (very weak, weak, reasonable, strong, very strong ).
I suggest leaving <code>switch</code> construction as it will make the code more readable in future versions.

File: lib/PasswordStrength/PasswordStrength.js (hunk 3 above)
Comment (oleksii-popov) [reply lib/PasswordStrength/PasswordStrength.js]: React render want to get something here. Refactored into <code>return null</code> and simplified.

File: lib/PasswordStrength/readme.md (hunk 4 above)
Comment (oleksii-popov) [reply lib/PasswordStrength/readme.md]: Yes, I forgot to update attribute name in the .md. Nice catch.

File: lib/PasswordStrength/readme.md (hunk 5 above)
Comment (oleksii-popov) [reply lib/PasswordStrength/readme.md]: It can be 'confirmed password' or 'change password' or 'new password'. Just want to avoid blind copy-paste. Developers should pay attention to what exactly they are working with.

File: lib/PasswordStrength/PasswordStrength.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+  render() {
+    if (!this.props.password) {
+      return null;
+    }
+
+    return (
+      <div aria-live="polite">
+        <FormattedMessage id={'stripes-smart-components.password-strength.' + this.output} />
+      </div>
```
Comment (JohnC-80) [inline lib/PasswordStrength/PasswordStrength.js]: It's best that the aria-live region is present already before a message is placed there... The early presence of the aria-live region is what will allow the assistive technology to monitor it for changes so that those can be announced. If the whole thing appears at once, it may not be picked up. You could have it render the empty `aria-live` div by default and then conditionally render the formatted message.

Comment (oleksii-popov) [reply lib/PasswordStrength/PasswordStrength.js]: Then I can remove <code>aria-live="polite"</code> here and and add it into readme.md example. Am I correct?
Also, can you please suggest how do you test accessibility? I used "Screen Reader" and it works properly.

Comment (oleksii-popov) [reply lib/PasswordStrength/PasswordStrength.js]: Updated

File: lib/PasswordStrength/readme.md (hunk 4 above)
Comment (zburke) [reply lib/PasswordStrength/readme.md]: Looks like this still needs to be fixed.

File: translations/stripes-smart-components/en.json [hunk 8]
```diff
@@ -82,5 +82,9 @@
   "deleteTag": "Remove tag",
   "enterATag": "Enter a tag.",
   "tagsTextArea": "Tag text area",
-  "addTag": "Add"
+  "addTag": "Add",
+
+  "password-strength.weak": "Your password is too weak.",
```
Comment (zburke) [inline translations/stripes-smart-components/en.json]: How about `passwordStrength.weak` etc to be consistent with the capitalization scheme elsewhere in this file?

Comment (oleksii-popov) [reply translations/stripes-smart-components/en.json]: Sure. Updated

File: lib/PasswordStrength/PasswordStrength.js (hunk 2 above)
Comment (oleksii-popov) [reply lib/PasswordStrength/PasswordStrength.js]: @MikeTaylor refactored into if/else.

**Comment by oleksii-popov** [comment]
Looks like this piece of code approved. I would like to ask anyone, who has write access to this repository to merge PR.

**Comment by MikeTaylor** [comment]
Done. Thanks for your work on this.
